astropy/nddata/nduncertainty.py

```diff
@@ -425,230 +671,243 @@ def _convert_uncertainty(self, other_uncert):
-
-            # We want the result to have the same unit as the parent, so we
-            # only need to convert the unit of the other uncertainty if it is
-            # different from its data's unit.
-            if other_uncert.unit != other_uncert.parent_nddata.unit:
-                other = (other_uncert.array * other_uncert.unit).to(
-                            other_uncert.parent_nddata.unit).value
-            else:
-                other = other_uncert.array
-            return np.abs(self.parent_nddata.data * other)
+        >>> ndd.uncertainty.array = 4
+        >>> ndd.uncertainty
+        VarianceUncertainty(4)
 
-        elif other_uncert.array is None:
-            # Formula: sigma = dA * |B|
+    .. note::
+        The unit will not be displayed.
+    """
+    @property
+    def uncertainty_type(self):
+        """``"var"`` : `VarianceUncertainty` implements variance.
+        """
+        return 'var'
 
-            # Just the reversed case
-            if self.unit != self.parent_nddata.unit:
-                this = (self.array * self.unit).to(
-                                            self.parent_nddata.unit).value
-            else:
-                this = self.array
-            return np.abs(other_uncert.parent_nddata.data * this)
+    @property
+    def supports_correlated(self):
+        """`True` : `VarianceUncertainty` allows to propagate correlated \
+                    uncertainties.
 
-        else:
-            # Formula: sigma = |AB|*sqrt((dA/A)**2+(dB/B)**2+2*dA/A*dB/B*cor)
+        ``correlation`` must be given, this class does not implement computing
+        it by itself.
+        """
+        return True
 
-            # This formula is not very handy since it generates NaNs for every
-            # zero in A and B. So we rewrite it:
+    @property
+    def unit(self):
+        """`~astropy.units.Unit` : The unit of the uncertainty, if any.
 
-            # Formula: sigma = sqrt((dA*B)**2 + (dB*A)**2 + (2 * cor * ABdAdB))
+        Even though it is not enforced the unit should be convertible to the
+        ``parent_nddata`` unit. Otherwise uncertainty propagation might give
+        wrong results.
 
-            # Calculate: dA * B (left)
-            if self.unit != self.parent_nddata.unit:
-                # To get the unit right we need to convert the unit of
-                # each uncertainty to the same unit as it's parent
-                left = ((self.array * self.unit).to(
-                        self.parent_nddata.unit).value *
-                        other_uncert.parent_nddata.data)
+        If the unit is not set the square of the unit of the parent will
+        be returned.
+        """
+        if self._unit is None:
+            if (self._parent_nddata is None or
+                    self.parent_nddata.unit is None):
+                return None
             else:
-                left = self.array * other_uncert.parent_nddata.data
+                return self.parent_nddata.unit**2
```

Comment:
Shouldn't a warning be issued about using parent units? It might not be the desired choice.. It is also a good practice to enforce it.